---
title: Using escape to perform ssGSEA analyses on single-cell RNA-seq data
author: 
- name: Nick Borcherding
  email: ncborch@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
- name: Jared Andrews
  email: jared.andrews07@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA

date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

output:
  BiocStyle::html_document:
    toc_float: true
package: escape
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Escape-ingToWork}
  %\VignetteEncoding{UTF-8} 
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Loading Processed Single-Cell Data


For the demonstration of *escape*, we will use the example "pbmc_small" data from *Seurat* and also generate a `SingleCellExperiment` object from it.


```{r}
suppressPackageStartupMessages(library(escape))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(SeuratObject))
pbmc_small <- get("pbmc_small")
sce.pbmc <- as.SingleCellExperiment(pbmc_small, assay = "RNA")
```

# Getting Gene Sets

## Option 1: Molecular Signture Database

The first step in the process of performing gene set enrichment analysis is identifying the gene sets we would like to use. The function `getGeneSets()` allows users to isolate a whole or multiple libraries from a list of GSEABase GeneSetCollection objects. 

We can do this for gene set collections from the built-in [Molecular Signature Database](https://www.gsea-msigdb.org/gsea/msigdb/search.jsp) by setting the parameter **library** equal to library/libraries of interest. For multiple libraries, just set **library = c("Library1", "Library2", etc)**. 

In Addition:

* **subcategory** Can be used to select subcategories in the larger database.  
* **gene.sets** Can select individual pathways/gene sets to be isolated.  

If the sequencing of the single-cell data is performed on a species other than "Homo sapiens", make sure to use the **species** parameter in `getGeneSets()` in order to get the correct gene nomenclature.

```{r}
GS.hallmark <- getGeneSets(library = "H")
```

## Option 2: Built-In gene sets

#TODO Add more details here

```{r, eval = FALSE}
data("escape.gene.sets", package="escape")
gene.sets <- escape.gene.sets
```


## Option 3: Define personal gene sets

```{r, eval=FALSE, tidy=FALSE}
gene.sets <- list(Bcells = c("MS4A1","CD79B","CD79A","IGH1","IGH2")
			            Myeloid = c("SPI1","FCER1G","CSF1R"),
			            Tcells = c("CD3E", "CD3D", "CD3G", "CD7","CD8A"))
			            
```

# Perfomring Enrichment

#TODO Add infor on the methods

## escape.matrix

escape has 2 major functions - the first being ```escape.matrix()```, which is the serves as the backbone of enrichment calculations. Using count-level data supplied from a single-cell object or matrix, ```escape.matrix()``` will produce enrichment score for the individual cells with the gene sets selected and output the values as a matrix.

**method** 

* AUCell 
* GSVA 
* ssGSEA
* Ucell

**groups**  

* The number of cells to calculate at once. 

**min.size** 

* The minimum size of detectable genes in a gene set. Gene sets less than the **min.size** will be removed before the calculation.

**normalize**

* Use the number of genes from the gene sets in data to normalize the enrichment scores. The default value is **FALSE**.


```{r tidy = FALSE}
enrichment.scores <- escape.matrix(obj = pbmc_small, 
                                    gene.sets = GS.hallmark[1:2], 
                                    groups = 1000, 
                                    cores = 2, 
                                    min.size = 5)

ggplot(data = as.data.frame(enrichment.scores), 
      mapping = aes(enrichment.scores[,1], enrichment.scores[,2])) + 
  geom_point() + 
  theme_classic()
```

## runEscape

Alternatively, we can use ```runEscape()``` to calculate the enrichment score and directly attach the output to a single-cell object. Th additional parameter for ```runEscape` is **assay.name**, in order to save the enrichment scores as a custom assay in the single-cell object. 


```{r tidy = FALSE}
pbmc_small <- escape.matrix(obj = pbmc_small, 
                            method = "ssGSEA",
                            gene.sets = GS.hallmark, 
                            groups = 1000, 
                            min.size = 5,
                            assay.name = "escape.ssGSEA")

sce.pbmc <- escape.matrix(obj = pbmc_small, 
                          method = "UCell",
                          gene.sets = GS.hallmark, 
                          groups = 1000, 
                          min.size = 5,
                          assay.name = "escape.UCell")
```

****

# Visualizations



****

# Statistical Analysis




***

# Conclusions

If you have any questions, comments or suggestions, feel free to visit the [github repository](https://github.com/ncborcherding/escape) or [email me](mailto:ncborch@gmail.com).

```{r}
sessionInfo()
```
