---
title: Using escape to perform gene set enrichment analyses on single-cell RNA-seq data
author: 
- name: Nick Borcherding
  email: ncborch@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA
- name: Jared Andrews
  email: jared.andrews07@gmail.com
  affiliation: Washington University in St. Louis, School of Medicine, St. Louis, MO, USA

date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'

output:
  BiocStyle::html_document:
    toc_float: true
package: escape
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Escape-ingToWork}
  %\VignetteEncoding{UTF-8} 
---

```{r, echo=FALSE, results="hide", message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
library(BiocStyle)
```

# Loading Processed Single-Cell Data

For the demonstration of *escape*, we will use the example **"pbmc_small"** data from *Seurat* and also generate a `SingleCellExperiment` object from it.


```{r}
suppressPackageStartupMessages(library(escape))
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(scran))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(SeuratObject))
suppressPackageStartupMessages(library(RColorBrewer))


pbmc_small <- get("pbmc_small")
sce.pbmc <- as.SingleCellExperiment(pbmc_small, assay = "RNA")
```

# Getting Gene Sets

## Option 1: Molecular Signture Database

The first step in the process of performing gene set enrichment analysis is identifying the gene sets we would like to use. The function `getGeneSets()` allows users to isolate a whole or multiple libraries from a list of GSEABase GeneSetCollection objects. 

We can do this for gene set collections from the built-in [Molecular Signature Database](https://www.gsea-msigdb.org/gsea/msigdb/search.jsp) by setting the parameter **library** equal to library/libraries of interest. For multiple libraries, just set **library = c("Library1", "Library2", etc)**. 

Additional parameters include: 

* **subcategory** Can be used to select subcategories in the larger database.  
* **gene.sets** Can select individual pathways/gene sets to be isolated.  

If the sequencing of the single-cell data is performed on a species other than "Homo sapiens", make sure to use the **species** parameter in `getGeneSets()` in order to get the correct gene nomenclature.

```{r}
GS.hallmark <- getGeneSets(library = "H")
```

## Option 2: Built-In gene sets

#TODO Add more details here

```{r, eval = FALSE}
data("escape.gene.sets", package="escape")
gene.sets <- escape.gene.sets
```

## Option 3: Define personal gene sets

```{r, eval=FALSE, tidy=FALSE}
gene.sets <- list(Bcells = c("MS4A1","CD79B","CD79A","IGH1","IGH2")
			            Myeloid = c("SPI1","FCER1G","CSF1R"),
			            Tcells = c("CD3E", "CD3D", "CD3G", "CD7","CD8A"))
```

# Perfomring Enrichment Calculation

#TODO Add infor on the methods

## escape.matrix

escape has 2 major functions - the first being ```escape.matrix()```, which is the serves as the backbone of enrichment calculations. Using count-level data supplied from a single-cell object or matrix, ```escape.matrix()``` will produce enrichment score for the individual cells with the gene sets selected and output the values as a matrix.

**method** 

* AUCell 
* GSVA 
* ssGSEA
* Ucell

**groups**  

* The number of cells to calculate at once. 

**min.size** 

* The minimum size of detectable genes in a gene set. Gene sets less than the **min.size** will be removed before the calculation.

**normalize**

* Use the number of genes from the gene sets in data to normalize the enrichment scores. The default value is **FALSE**.


```{r tidy = FALSE}
enrichment.scores <- escape.matrix(pbmc_small, 
                                   gene.sets = GS.hallmark, 
                                   groups = 1000, 
                                   min.size = 5)

ggplot(data = as.data.frame(enrichment.scores), 
      mapping = aes(enrichment.scores[,1], enrichment.scores[,2])) + 
  geom_point() + 
  theme_classic() + 
  theme(axis.title = element_blank())
```

## runEscape

Alternatively, we can use ```runEscape()``` to calculate the enrichment score and directly attach the output to a single-cell object. Th additional parameter for ```runEscape` is **new.assay.name**, in order to save the enrichment scores as a custom assay in the single-cell object. 


```{r tidy = FALSE}
pbmc_small <- runEscape(pbmc_small, 
                        method = "ssGSEA",
                        gene.sets = GS.hallmark, 
                        groups = 1000, 
                        min.size = 5,
                        new.assay.name = "escape.ssGSEA")

sce.pbmc <- runEscape(sce.pbmc, 
                      method = "UCell",
                      gene.sets = GS.hallmark, 
                      groups = 1000, 
                      min.size = 5,
                      new.assay.name = "escape.UCell")
```

****

# Visualizations

There are a number of ways to look at the enrichment values downstream of ```runEscape()``` with the myriad plotting and visualizations functions/packages for single-cell data. *escape* include several additional plotting functions to assist in the analysis.

## heatmapEnrichment

We can examine the enrichment values across our gene sets by using ```heatmapEnrichment()```. This visualization will return the mean of the **group.by** variable. As a default - all visualizations of single-cell objects will use the cluster assignment or active identity as a default for visualizations.

```{r}
heatmapEnrichment(pbmc_small, 
                  group.by = "ident",
                  gene.set.use = "all",
                  assay = "escape.ssGSEA")
```
Most of the visualizations in *escape* have a defined set of parameters.

**group.by**

* The grouping variable for the comparison.

**facet.by**

* Using a variable to facet the graph into seperate visualizations.

**scale**

* **TRUE** - z-trasnform the enrichment values.
* **FALSE** - leave raw values (**DEFAULT**).

In addition, ```heatmapEnrichment()``` allows for the reclustering of rows and columns using Euclidean distance of the enrichment scores and the Ward2 methods for clustering using **cluster.rows** and **cluster.columns**.

```{r}
heatmapEnrichment(sce.pbmc, 
                  group.by = "ident",
                  assay = "escape.UCell",
                  scale = TRUE,
                  cluster.rows = TRUE,
                  cluster.columns = TRUE)
```

Each visualization has an additional argument called **palette that supplies the coloring scheme to be used - available color palettes can be viewed with ```hcl.pals()```. 

```{r}
hcl.pals()
```

```{r}
heatmapEnrichment(pbmc_small, 
                  assay = "escape.ssGSEA",
                  palette = "Spectral") 
```

Alternatively, we can add an additional layer to the ggplot object that is returned by the visualizations using something like ```scale_fill_gradientn()``` for continuous values or ```scale_fill_manual()``` for the categorical variables.

```{r}
heatmapEnrichment(sce.pbmc, 
                  group.by = "ident",
                  assay = "escape.UCell") + 
  scale_fill_gradientn(colors = rev(brewer.pal(11, "RdYlBu"))) 
```

## geyserEnrichment

We can also focus on individual gene sets - one approach is to use ```geyserEnrichment()```. Here individual cells are plotted along the Y-axis with graphical summary where the central dot refers to the median enrichment value and the thicker/thinner lines demonstrate the interval summaries referring the 66% and 95%.

```{r}
geyserEnrichment(pbmc_small, 
                 assay = "escape.ssGSEA",
                 gene.set = "HALLMARK-INTERFERON-GAMMA-RESPONSE")
```

To show the additional parameters that appear in visualizations of individual enrichment gene sets - we can reorder the groups by the mean of the gene set using **order.by** = "mean".

```{r}
geyserEnrichment(pbmc_small, 
                 assay = "escape.ssGSEA",
                 gene.set = "HALLMARK-INTERFERON-GAMMA-RESPONSE", 
                 order.by = "mean")
```

What if we had 2 separate samples or groups within the data? Another parameter we can use is **facet.by** to allow for direct visualization of an additional variable. 

```{r}
geyserEnrichment(pbmc_small, 
                 assay = "escape.ssGSEA",
                 gene.set = "HALLMARK-INTERFERON-GAMMA-RESPONSE", 
                 facet.by = "groups")
```
Lastly, we can select the way the color is applied to the plot using the **color.by** parameter. Here we can set it to the gene set of interest *"HALLMARK-INTERFERON-GAMMA-RESPONSE"*.

```{r}
geyserEnrichment(pbmc_small, 
                 assay = "escape.ssGSEA",
                 gene.set = "HALLMARK-INTERFERON-GAMMA-RESPONSE", 
                 color.by  = "HALLMARK-INTERFERON-GAMMA-RESPONSE")
```

## ridgeEnrichment

Similar to the ```geyserEnrichment()``` the ```ridgeEnrichment()``` can display the distribution of enrichment values across the selected gene set. The central line is at the median value for the respective grouping. 

```{r}
ridgeEnrichment(sce.pbmc, 
                assay = "escape.UCell",
                gene.set = "HALLMARK-IL2-STAT5-SIGNALING")
```
We can get the relative position of individual cells along the x-axis using the **add.rug** parameter.

```{r}
ridgeEnrichment(sce.pbmc, 
                assay = "escape.UCell",
                gene.set = "HALLMARK-IL2-STAT5-SIGNALING",
                add.rug = TRUE)
```

## splitEnrichment

Another distribution visualization is a violin plot, which we separate and directly compare using a binary classification. Like ```ridgeEnrichment()```, this allows for greater use of categorical variables. For ```splitEnrichment()```, the output will be two halves of a violin plot based on the **split.by** parameter with a central boxplot with the relative distribution across all samples.

```{r}
splitEnrichment(pbmc_small, 
                assay = "escape.ssGSEA",
                gene.set = "HALLMARK-IL2-STAT5-SIGNALING", 
                split.by = "groups")
```

## densityEnrichment

```densityEnrichment()``` is a method to visualize the mean rank position of the gene set features along the total feature space by group. This is similar to traditional GSEA analysis, but is not calculating the walk-based enrichment score.

**gene.set.use** 

* The selected gene set to visualize

**gene.set.reference**

* The gene set library from either any of the 3 options in the first section of the vignette.

```{r}
densityEnrichment(pbmc_small, 
                  gene.set.use = "HALLMARK_IL6_JAK_STAT3_SIGNALING", 
                  gene.set.reference = GS.hallmark)
```
## scatterEnrichment

It may be advantageous to look at the distribution of multiple gene sets - here we can use ```scatterEnrichment()``` for a 2 gene set comparison. The color values are based on the density of points determine by the number of neighbors, similar to the [Nebulosa R package](https://www.bioconductor.org/packages/release/bioc/html/Nebulosa.html). We just need to define which gene set to plot on the **x.axis** and which to plot on the **y.axis**.


```{r}
scatterEnrichment(sce.pbmc, 
                  assay = "escape.UCell",
                  x.axis = "HALLMARK-INTERFERON-GAMMA-RESPONSE",
                  y.axis = "HALLMARK-IL6-JAK-STAT3-SIGNALING")
```
The scatter plot can also be converted into a hexbin, another method for summarizing the individual cell distributions along the x and y axis, by setting **style** = "hex". 

```{r}
scatterEnrichment(sce.pbmc, 
                  assay = "escape.UCell",
                  x.axis = "HALLMARK-INTERFERON-GAMMA-RESPONSE",
                  y.axis = "HALLMARK-IL6-JAK-STAT3-SIGNALING", 
                  style = "hex")
```

****

# Statistical Analysis

## Pricipal Component Analysis (PCA)

escape has its own PCA function ```performPCA()``` which will work on a single-cell object or a matrix of enrichment values. This is specifically useful for downstream visualizations as it stores the eigen values and rotations. If we want to look at the the relative contribution to overall variance of each component or a Biplot-like overlay of the individual features, use ```performPCA()```.

Alternatively, other PCA-based functions like Seurat's ```RunPCA()``` or scater's ```runPCA()` can be used. These functions are likely faster and would be ideal if we have a larger number of cells and/or gene sets.

```{r}
pbmc_small <- performPCA(pbmc_small, 
                         assay = "escape.ssGSEA",
                         n.dim = 1:10)
```




## Differential Enrichment


***

# Conclusions

If you have any questions, comments or suggestions, feel free to visit the [github repository](https://github.com/ncborcherding/escape) or [email me](mailto:ncborch@gmail.com).

```{r}
sessionInfo()
```
